cfgmgr/vrfmgr.cpp

```diff
@@ -0,0 +1,146 @@
+#include <string.h>
+#include "logger.h"
+#include "dbconnector.h"
+#include "producerstatetable.h"
+#include "tokenize.h"
+#include "ipprefix.h"
+#include "vrfmgr.h"
+#include "exec.h"
+#include "shellcmd.h"
+
+#define VRF_TABLE_START 100
+#define VRF_TABLE_END 164
+
+using namespace swss;
+
+VrfMgr::VrfMgr(DBConnector *cfgDb, DBConnector *appDb, DBConnector *stateDb, const vector<string> &tableNames) :
+        Orch(cfgDb, tableNames)
+{
+    for (uint32_t i = VRF_TABLE_START; i < VRF_TABLE_END; i++)
+    {
+        m_freeTables.emplace(i);
+    }
+}
+
+uint32_t VrfMgr::getFreeTable(void)
+{
+    SWSS_LOG_ENTER();
+
+    if (m_freeTables.empty())
+    {
+        return 0;
+    }
+
+    uint32_t table = *m_freeTables.begin();
+    m_freeTables.erase(table);
+
+    return table;
+}
+
+void VrfMgr::returnTable(uint32_t table)
+{
+    SWSS_LOG_ENTER();
+
+    m_freeTables.emplace(table);
+}
+
+bool VrfMgr::delLink(const string& vrfName)
+{
+    SWSS_LOG_ENTER();
+
+    stringstream cmd;
+    string res;
+
+    if (m_vrfTableMap.find(vrfName) == m_vrfTableMap.end())
+    {
+        return false;
+    }
+
+    returnTable(m_vrfTableMap[vrfName]);
+
+    cmd << IP_CMD << " link del " << vrfName;
+    int ret = swss::exec(cmd.str(), res);
+    if (ret)
+    {
+        SWSS_LOG_ERROR("Failed command %s, rc:%d", cmd.str().c_str(), ret);
+        return false;
+    }
+
+    m_vrfTableMap.erase(vrfName);
+
+    return true;
+}
+
+bool VrfMgr::setLink(const string& vrfName)
+{
+    SWSS_LOG_ENTER();
+
+    stringstream cmd;
+    string res;
+
+    uint32_t table = getFreeTable();
+    if (table == 0)
+    {
+        return false;
+    }
+
+    cmd << IP_CMD << " link add " << vrfName << " type vrf table " << table;
+    int ret = swss::exec(cmd.str(), res);
```

Comment:
I'm think about warm restart support for vrfmgrd.   If the vrf table has been configured on linux,  adding it again would return error.  Could it be changed to warm restart agnostic, like checking the existence of vrf table before adding it?
```
root@sonic:/home/admin# ip link add vrftest1 type vrf table 120
root@sonic:/home/admin# ip link add vrftest1 type vrf table 120
RTNETLINK answers: File exists
root@sonic:/home/admin# echo $?
2
```

Reply:
Not sure if I follow. Warm restart reboots kernel. VRF netdevs will be gone after reboot.

Reply:
There is SWSS docker warm restart. In that case, Linux env is kept.

Reply:
there is no problem to get all vrfs using following command. 
```
admin@str-s6000-on-4:~$ ip -br link show type vrf
vrftest1         DOWN           c2:31:e9:01:c8:f4 <NOARP,MASTER> 
```

the problem is to get the route table id.

Reply:
do we need the table id? I can see the ip route add can use vrfname directly. @marian-pritsak , do you know who is going to use that table id?

Reply:
from what I can see, it seems we can get form netlink messages.

https://github.com/FRRouting/frr/blob/master/zebra/if_netlink.c#L305

I do not know if there is a command line that can get this.

Reply:
Done

Reply:
ip -d shows table ID

Reply:
thanks